### PR TITLE
MONGOSH-106, MONGOSH-107: add telemetry opt-in command, add userid

### PR DESCRIPTION
**Pull request by lrlna**
## Description
This PR creates a config file where currently two things are stored:
- userId
- telemetry opt-in.

For example:
Capture d’écran, le 2020-04-01 à 17 18 35

When user initial logs in, we write the file, setup `userId`, and toggle `enableTelemetry` to false. User also receives a message to opt in. They can subsequently opt in with `enableTelemetry` command or opt out with `disableTelemetry`.

For example: 
Capture d’écran, le 2020-04-01 à 17 15 21

## Next Steps
This is not yet hooked up to Segment, so next thing would be to do that, as well as hook up the logger to be notified when user opts in and opts out. That way it can take into consideration when to write to Segment.

### CODE REVIEW

File: packages/cli-repl/src/cli-repl.ts [hunk 1]
```diff
@@ -68,6 +75,48 @@ class CliRepl {
...
+  }
+
+  /**
+  * Checks if config file exists.
+  *
+  * If exists: sets userId and enabledTelemetry to this.
+  * If does not exist: writes a new file with a newly generated ObjectID for
+  * userid and enableTelemetry set to false.
+  */
+  generateOrReadTelemetryConfig(): void {
```
Comment (mcasimir) [inline packages/cli-repl/src/cli-repl.ts:98]: How do we sync `fs.open` here?

File: packages/cli-repl/src/cli-repl.ts [hunk 2]
```diff
@@ -197,6 +239,28 @@ class CliRepl {
...
+  *
+  * @returns {string} Status of telemetry logging: disabled/enabled
+  */
+  toggleTelemetry(enabled: boolean): string {
+    this.enableTelemetry = enabled;
+
+    const configPath = path.join(this.mongoshDir, 'config');
+    const config = { userId: this.userId, enableTelemetry: this.enableTelemetry };
+    fs.writeFileSync(configPath, JSON.stringify(config));
+
```
Comment (mcasimir) [inline packages/cli-repl/src/cli-repl.ts:256]: These line are also in `generateOrReadTelemetryConfig` would it be an idea to put them together in a function? Dunno writeConfig, readConfig :/

File: packages/cli-repl/src/cli-repl.ts (hunk 1 above)
Comment (mcasimir) [reply packages/cli-repl/src/cli-repl.ts:98]: like if `fs.open` would take long enough (or it would hang)

**Comment by mcasimir** [comment]
Noticed now bootstrap fails too

File: packages/cli-repl/src/constants.ts [hunk 3]
```diff
@@ -1,6 +1,12 @@
 import i18n from '@mongosh/i18n';
 import clr from './clr';
 
+export const TELEMETRY = `
+${i18n.__('cli-repl.cli-repl.telemetry')}
+${i18n.__('cli-repl.cli-repl.enableTelemetry')}${clr('enableTelemtry()', 'bold')}.
```
Comment (mcasimir) [inline packages/cli-repl/src/constants.ts:6]: I think your worst nightmare is here again :D

File: packages/cli-repl/src/cli-repl.ts (hunk 1 above)
Comment (lrlna) [reply packages/cli-repl/src/cli-repl.ts:98]: Both `fs.open` and `fs.openSync` could be blocking or take long enough, the only difference (iirc) is that `fs.open` is backed by a thread pool. If you are worried about it taking too long, I can look into how to add a timeout. 

I did look at having it done synchronously, and here is what that looked like:

```js
    let fd;

    try {
      fd = fs.openSync(configPath, 'wx');

      this.userId = new ObjectId(Date.now());
      this.enableTelemetry = false;
      const config = { userId: this.userId, enableTelemetry: this.enableTelemetry };
      fs.writeFileSync(fd, JSON.stringify(config));
    } catch (err) {
      if (err.code === 'EEXIST') {
        const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
        this.userId = config.userId;
        this.enableTelemetry = config.enableTelemetry;
        return;
      }
      this.bus.emit('error', e)
      throw err;

    } finally {
      if (fd !== undefined) fs.closeSync(fd);
    }
```

File: packages/cli-repl/src/cli-repl.ts (hunk 2 above)
Comment (lrlna) [reply packages/cli-repl/src/cli-repl.ts:256]: Good catch! I didn't even notice 😅 Will write a lil abstracted function

File: packages/cli-repl/src/constants.ts (hunk 3 above)
Comment (lrlna) [reply packages/cli-repl/src/constants.ts:6]: it took me literally 20 min to understand what you meant :D

File: packages/cli-repl/src/cli-repl.ts (hunk 1 above)
Comment (mcasimir) [reply packages/cli-repl/src/cli-repl.ts:98]: I meant that with the async version we could have race conditions. Like `this.userId not being set when we try to read it.

Comment (lrlna) [reply packages/cli-repl/src/cli-repl.ts:98]: a race condition is a horse of a different colour! i can make it sync.

**Comment by lrlna** [comment]
@mcasimir I think I addressed everything you've mentioned now.

In terms of testing -- it'll have to be under end to end tests. I don't think there is a way to do this otherwise :/. I have so far thought of possibly moving all the reads and writes over to another file, but I feel like that will give us only the ability to test actual `fs` functions (which is already done in node itself), and creates a bit more work around managing and passing around information such as `this.userId`, or `this.mongoshDir` or `this.enableTelemetry`.

**Comment by mcasimir** [comment]
Yeah i think we are good with e2e tests, let's add some close after then.

> I have so far thought of possibly moving all the reads and writes over to another file, but I feel like that will give us only the ability to test actual `fs` functions.

Was not suggesting that .. but it doesn't sound like a bad idea to me. If we could do constructor injection (`new CliRepl(dep1, dep2, dep3)`) it would be just great and improve the testability, but, even without we could get a better separation of concerns: `CliRepl` would know to use the configuration but not how to store, retrieve, serialize and deserialize it, so it would have one responsiblity less.

**Comment by mcasimir** [comment]
also LGTM :)

**Pull request merged**